### DISCUSSION TITLE
Attach Characters

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#3244] [Development]
I propose a new config variable: `$attach_chars`.
They'd allow the user to customise the display of attachments.

Here are some suggested icons along with the current text representation:

| Text  | Icon  | Description       |
| :---: | :---: | :---------------- |
| `I`   | `📄`  | inline            |
| `A`   | `📎`  | attachment        |
| `Q`   | `⑫`  | countable         |
| `*`   | `✅`  | tagged            |
| `c`   | `🔄`  | convert           |
| `d`   | `❌`  | deleted           |
| `-`   | `⏳`  | unlink            |
| none  | `👪`  | multipart related |
| none  | `🔀`  | multipart mixed   |
| none  | `🌍`  | multipart lingual |

e.g. `set attach_chars = "📄📎⑫✅🔄❌⏳👪🔀🌍"`

**See also**:

- [`$crypt_chars`](https://neomutt.org/guide/reference#crypt-chars) - User-configurable crypto flags: signed, encrypted etc.
- [`$flag_chars`](https://neomutt.org/guide/reference#flag-chars) - User-configurable index flags: tagged, new, etc
- [`$from_chars`](https://neomutt.org/guide/reference#from-chars) - User-configurable index flags: to address, cc address, etc
- [`$status_chars`](https://neomutt.org/guide/reference#status-chars) - Indicator characters for the status bar
- [`$to_chars`](https://neomutt.org/guide/reference#to-chars) - Indicator characters for the 'To' field in the index

**Comment by flatcap** [comment]
We might also want icons for:
- `multipart/alternative` → :left_right_arrow: 
- `multipart/signed` → :lock_with_ink_pen:
- `multipart/encrypted` → :lock: 
- `message/rfc822` → :envelope: 
- An attached PGP key → :closed_lock_with_key:

**Reply by dcpurton** [reply]
`application/pgp-keys` is in a slightly different category because it's a particular sub-type of the `application` major MIME type.

This opens up a whole different world of potentially associating icons with arbitrary MIME types.

**Comment by dcpurton** [comment]
You've suggested icons for various `multipart` attachments. But presumably you could just as easily put icons in the same column for other major MIME types:

- `TYPE_OTHER`: :question: 
- `TYPE_AUDIO`: :loud_sound: 
- `TYPE_APPLICATION`: :gear: 
- `TYPE_IMAGE`: :framed_picture: 
- `TYPE_MESSAGE`: :envelope: 
- `TYPE_MODEL`: :nut_and_bolt: 
- `TYPE_MULTIPART`: :houses: (generic form for sub-types not specifically covered)
- `TYPE_TEXT`: :page_facing_up: 
- `TYPE_VIDEO`: :movie_camera: 
- `TYPE_ANY`: :asterisk: 

There's also `font`: :capital_abcd: (https://www.iana.org/assignments/media-types/media-types.xhtml#font)

For reasons I can't quite explain there's no `TYPE_FONT` in `email/mime.h`, but Neomutt seems to support it just fine. (I guess from `/etc/mime.types`)

**Comment by flatcap** [comment]
That's a lot of icons.
I made a mistake in my initial post -- mixing in the mime types.

The [`$from_chars`](https://neomutt.org/guide/reference#from-chars) style of icons works well for a small set of fixed attributes.
Let's consider a new `$attach_chars` for just them.

Currently, the default [`$attach_format`](https://neomutt.org/guide/reference#attach-format) has four separate expandos:
- `%u` - Unlink
- `%D` - Deleted flag
- `%I` - I (inline) or A (attachment)
- `%t` - Tagged flag

Add in a couple of lesser-used expandos:
- `%Q` - countable
- `%c` - convert

I think we can create an general-purpose **three**-column flags field, like `%Z` in the [`$index_format`](https://neomutt.org/guide/reference#index-format)
Consider:

| Column | Text | Expando | Icon | Description | Used in |
| :----: | :--- | :------ | :--- | :---------- | :------ |
| 1      | `-`  | `%u`    | `⏳` | unlink      | Compose |
| 1      | `d`  | `%d`    | `❌` | deleted     | Receive |
|        |      |         |      |             |         |
| 2      | `I`  | `%I`    | `📄` | inline      | Both    |
| 2      | `A`  | `%I`    | `📎` | attachment  | Both    |
|        |      |         |      |             |         |
| 3      | `*`  | `%t`    | `✅` | tagged      | Both    |
| 3      | `Q`  | `%Q`    | `⑫`  | countable   | Both    |
| 3      | `c`  | `%c`    | `🔄` | convert     | Compose |

Unlink/deleted aren't used together.
Inline/attachment are mutually exclusive.
Tagged would have higher-priority than countable/convert.

**Reply by dcpurton** [reply]
Yes, this is sensible I think.

Does `ncurses` deal with these double width characters easily?

**Reply by flatcap** [reply]
Yeah, ncurses is fine and we can measure the char-width in the code (we already do, elsewhere)

**Comment by flatcap** [comment]
All those icons for mime-types look great, especially placed after `%T` (tree characters).

```
↔️ <no description>                            [multipart/alternative]
├─>📄 <no description>                                    [text/plain]
└─>🌍 <no description>                                     [text/html]
🖼️ cat.gif                                                 [image/gif]
📹 dog.avi                                               [image/video]
```

The question is, how do we configure them all?

**Reply by dcpurton** [reply]
Pretty indeed.

In the above example you're using both major and minor MIME type (both for `multipart` and `text`). This makes it harder than just dealing with major MIME types (of which there are not many) and you could easily configure with:

```
set_mime_chars = '❓🔊⚙️🔠🖼️✉️🔩🏘️📄🎞️*️⃣'
```

I think `dog.avi` would actually be: `video/x-msvideo`. At least that is what my `/etc/mime.types` gives.

**Reply by dcpurton** [reply]
As soon as you want to add in the ability to distinguish between minor MIME types, configuration becomes more challenging.

You might need a whole new thing analogous to colours:

```
icon mimetype 'multipart/alternative' '↔️'
icon mimetype 'multipart/related' '👪'
icon mimetype 'multipart/mixed' '🔀'
icon mimetype 'multipart/multilingual' '🌍'
```

**Reply by dcpurton** [reply]
hmm. Might need a better icon than `↔️` These arrows are commonly in fonts and I can't figure out how to override things with an emoji font… Plus you might not always want an emoji for this character.

**Reply by flatcap** [reply]
> I can't figure out how to override things with an emoji font

I never understood how they worked, either, but I've just done a bit of reading:
tl;dr (for me on Fedora)

<kbd>Ctrl-.</kbd> (ctrl-dot) gives me an emoji prompt (an underlined `e`).
Typing `arrow` + <kbd>\<space\></kbd> brings up an emoji matching "arrow"
Pressing <kbd>\<space\></kbd> more, cycles through the options (with a menu).

There's also a raw unicode prompt <kbd>Ctrl-Shift-U</kbd> (ctrl-capital-u)
Typing <kbd>2194</kbd> + <kbd>\<enter\></kbd> gives me: `↔`

If I then <kbd>Ctrl-Shift-U</kbd> again and type <kbd>fe0f</kbd> + <kbd>\<enter\></kbd>
that generates `[Variation Selector-16]` and the screen now shows: `↔️`
The VS-16 selects the emoji version of the previous character.

Refs:
- https://www.codejam.info/2021/11/emoji-variation-selector.html
- https://askubuntu.com/questions/1372781/new-ctrl-period-key-sequence-default-in-21-10
- https://unicode-table.com/en/

**Reply by dcpurton** [reply]
`↔️` and `↔`

**Reply by flatcap** [reply]
> You might need a whole new thing analogous to colours:
> `icon mimetype 'multipart/alternative' '↔️'`

An "icon" command with subtype "mimetype" -- very nice, very extensible

And we can throw in a regex too.
They'd be useful for default cases, catchalls.
